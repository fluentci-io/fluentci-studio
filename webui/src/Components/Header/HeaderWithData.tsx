import { FC, useEffect, useState } from "react";
import Header from "./Header";
import { useNavigate, useParams, useLocation } from "react-router-dom";
import {
  Project,
  Repository,
  Run,
  useGetActionsLazyQuery,
  useGetActionsQuery,
  useGetLinkedRepositoryLazyQuery,
  useGetProjectLazyQuery,
  useGetRunLazyQuery,
  useRunPipelineMutation,
} from "../../Hooks/GraphQL";
import { useRecoilState, useRecoilValue } from "recoil";
import { ComposerState } from "../../Containers/Project/MainContent/Composer/ComposerState";
import { AuthState } from "../../Containers/Auth/AuthState";
import { ProjectState } from "../../Containers/Project/ProjectState";

export type HeaderWithDataProps = {
  breadcrumbs?: { title: string; link?: string }[];
};

const HeaderWithData: FC<HeaderWithDataProps> = () => {
  const me = useRecoilValue(AuthState);
  const composerState = useRecoilValue(ComposerState);
  const navigate = useNavigate();
  const { pathname } = useLocation();
  const { id } = useParams();
  const [{ project }, setProjectState] = useRecoilState(ProjectState);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const [actions, setActions] = useState<any[]>([]);
  const [run, setRun] = useState<Run | null | undefined>(null);
  const [getProject] = useGetProjectLazyQuery({
    variables: {
      id: id!,
    },
    fetchPolicy: "cache-and-network",
  });
  const [getRun] = useGetRunLazyQuery({
    variables: {
      id: id!,
    },
  });
  const [runPipeline] = useRunPipelineMutation();
  const [getActions] = useGetActionsLazyQuery();
  const { data } = useGetActionsQuery({
    variables: {
      projectId: project?.id || "",
    },
    fetchPolicy: "cache-and-network",
  });
  const [linkedRepository, setLinkedRepository] = useState<
    Repository | null | undefined
  >(null);
  const [getLinkedRepository] = useGetLinkedRepositoryLazyQuery();
  const [loading, setLoading] = useState(false);

  const setProject = (value?: Project | null) =>
    setProjectState({ project: value });

  useEffect(() => {
    if (pathname.startsWith("/run")) {
      getRun().then(({ data }) => {
        setRun(data?.getRun);
        if (data?.getRun?.projectId) {
          getProject({
            variables: {
              id: data.getRun.projectId,
            },
            fetchPolicy: "cache-and-network",
          }).then(({ data }) => setProject(data?.project));
        }
      });
      return;
    }
    if (
      pathname.startsWith("/project") ||
      pathname.startsWith("/link-project")
    ) {
      getProject({
        variables: {
          id: id!,
        },
        fetchPolicy: "cache-and-network",
      }).then(({ data }) => setProject(data?.project));
      return;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pathname]);

  useEffect(() => {
    if (!me || !project) {
      return;
    }
    getLinkedRepository({
      variables: {
        projectId: project.id,
      },
      fetchPolicy: "cache-and-network",
    }).then((res) => setLinkedRepository(res.data?.linkedRepository));
  }, [me, getLinkedRepository, project]);

  useEffect(() => {
    if (!data) {
      return;
    }
    setActions(data.actions || []);
  }, [data]);

  const onRun = () => {
    runPipeline({
      variables: {
        projectId: project!.id!,
      },
    }).then(({ data }) => {
      setLoading(false);
      if (data?.runPipeline) {
        navigate(`/run/${data.runPipeline.id}`);
      }
    });
    setLoading(true);
  };

  const breadcrumbs: { title: string; link?: string }[] = [
    {
      title: "Projects",
      link: "/",
    },
    {
      title: project?.displayName || project?.name || "",
      link: run && project?.id ? `/project/${project.id}` : undefined,
    },
  ];

  if (run) {
    breadcrumbs.push({
      title: run.name,
    });
  }

  useEffect(() => {
    if (!project) {
      return;
    }
    setTimeout(() => {
      getActions({
        variables: {
          projectId: project.id,
        },
        fetchPolicy: "cache-and-network",
      }).then(({ data }) => {
        setActions(data?.actions || []);
      });
    }, 1500);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [composerState]);

  return (
    <Header
      id="1"
      onRun={onRun}
      breadcrumbs={breadcrumbs}
      showRunButton={
        !!project &&
        !pathname.startsWith("/link-project") &&
        (!!actions?.filter((x) => x.enabled).length ||
          (project?.path !== "empty" && !linkedRepository) ||
          (!!linkedRepository && !!actions?.filter((x) => x.enabled).length))
      }
      loading={loading}
      linkedRepository={linkedRepository}
    />
  );
};

export default HeaderWithData;
