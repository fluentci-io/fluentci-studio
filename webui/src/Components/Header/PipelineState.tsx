import { atom } from "recoil";

export const PipelineState = atom<{
  runs: { [key: string]: boolean | undefined };
}>({
  key: "PipelineState",
  default: {
    runs: {},
  },
});
