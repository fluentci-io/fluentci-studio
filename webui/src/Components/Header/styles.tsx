import styled from "@emotion/styled";
import { Link as DefaultLink } from "react-router-dom";

export const Link = styled(DefaultLink)`
  color: #fff;
  &:hover {
    color: #24ffb5 !important;
  }
`;

export const Title = styled.div`
  font-size: 25px;
  margin-bottom: 20px;
  font-weight: 500;
  flex: 1;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 20px;
`;

export const RunButton = styled.button`
  height: 40px;
  background-color: #24ffb5;
  color: #000;
  border: none;
  font-weight: 600;
  width: 150px;
  cursor: pointer;
  &:hover {
    background-color: #18d193;
  }
`;

export const PopoverButton = styled.button`
  width: 40px;
  height: 40px;
  cursor: pointer;
  background-color: initial;
  border: none;
  color: #fff;
  margin-right: 10px;
`;

export default {
  Tab: {
    Tab: {
      style: ({
        $isActive,
        $theme,
      }: {
        $isActive: string;
        $theme: { primaryFontFamily: string };
      }) => ({
        backgroundColor: "#0f0124",
        color: $isActive ? "#24ffb5" : "#fff",
        fontFamily: $theme.primaryFontFamily,
        fontSize: "16px",
        ":hover": {
          color: "#24ffb5",
          backgroundColor: "#0f0124",
        },
      }),
    },
  },
  Tabs: {
    TabHighlight: {
      style: {
        backgroundColor: "#24ffb5",
      },
    },
    TabBorder: {
      style: {
        height: "0px",
      },
    },
  },
  StatefulMenu: {
    Option: {
      props: {
        overrides: {
          ListItem: {
            style: ({ $theme }: { $theme: { primaryFontFamily: string } }) => ({
              fontFamily: $theme.primaryFontFamily,
            }),
          },
        },
      },
    },
    List: {
      style: {
        border: "1px solid rgb(71 5 94 / 29%)",
      },
    },
  },
};
