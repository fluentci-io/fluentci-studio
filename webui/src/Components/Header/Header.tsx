import { FC } from "react";
import { EllipsisVertical } from "@styled-icons/fa-solid";
import { StatefulPopover } from "baseui/popover";
import { StatefulMenu } from "baseui/menu";
import { Breadcrumbs } from "baseui/breadcrumbs";
import styles, { Link, Container, RunButton, PopoverButton } from "./styles";

export type HeaderProps = {
  id: string;
  onRun: (id: string) => void;
  menu?: string[];
  breadcrumbs?: { title: string; link?: string }[];
};

const Header: FC<HeaderProps> = (props) => {
  const { id, breadcrumbs, onRun, menu } = props;
  return (
    <Container>
      <Breadcrumbs
        overrides={{
          Root: {
            style: {
              display: "flex",
              flexDirection: "row",
              alignItems: "center",
            },
          },
          ListItem: {
            style: ({ $theme }) => ({
              fontFamily: $theme.primaryFontFamily,
            }),
          },
        }}
      >
        {breadcrumbs?.map(({ title, link }) => (
          <>
            {link && (
              <Link to={link} style={{ color: "#ffffffa8" }}>
                {title}
              </Link>
            )}
            {!link && <span>{title}</span>}
          </>
        ))}
      </Breadcrumbs>
      {!!menu?.length && (
        <StatefulPopover
          placement="bottomRight"
          content={({ close }) => (
            <StatefulMenu
              items={menu.map((label) => ({ label }))}
              onItemSelect={() => {
                close();
              }}
              overrides={styles.StatefulMenu}
            />
          )}
        >
          <PopoverButton onClick={(e) => e.stopPropagation()}>
            <EllipsisVertical size={20} />
          </PopoverButton>
        </StatefulPopover>
      )}
      <RunButton onClick={() => onRun(id)}>Run</RunButton>
    </Container>
  );
};

Header.defaultProps = {
  menu: [],
  breadcrumbs: [
    {
      title: "Projects",
      link: "/",
    },
  ],
};

export default Header;
