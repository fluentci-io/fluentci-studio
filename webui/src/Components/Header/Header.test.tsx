import { render } from "@testing-library/react";
import Header from "./Header";
import { BrowserRouter } from "react-router-dom";

describe("Header", () => {
  it("should render the component", () => {
    const { container } = render(
      <Header
        id="1"
        onRun={vi.fn()}
        breadcrumbs={[
          { title: "Projects", link: "/" },
          {
            title: "frosty-hamilton",
            link: "/project/yvzb2a5aaceuaa8vs9kawruc",
          },
        ]}
      />,
      {
        wrapper: BrowserRouter,
      }
    );
    expect(container).toMatchSnapshot();
  });
});
