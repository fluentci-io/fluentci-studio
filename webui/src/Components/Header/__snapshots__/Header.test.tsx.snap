// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should render the component 1`] = `
<div>
  <div
    class="css-rrj1n6"
  >
    <nav
      aria-label="Breadcrumbs navigation"
      class=""
      data-baseweb="breadcrumbs"
    >
      <ol
        class=""
      >
        <li
          class=""
        >
          <a
            class="css-lr01fi"
            href="/"
            style="color: rgba(255, 255, 255, 0.69);"
          >
            Projects
          </a>
          <div
            aria-hidden="true"
            class=""
          >
            <svg
              class=""
              data-baseweb="icon"
              title="Right"
              viewBox="0 0 24 24"
            >
              <title>
                Right
              </title>
              <path
                clip-rule="evenodd"
                d="M9.29289 7.29289C8.90237 7.68342 8.90237 8.31658 9.29289 8.70711L12.5858 12L9.29289 15.2929C8.90237 15.6834 8.90237 16.3166 9.29289 16.7071C9.68342 17.0976 10.3166 17.0976 10.7071 16.7071L14.7071 12.7071C14.8946 12.5196 15 12.2652 15 12C15 11.7348 14.8946 11.4804 14.7071 11.2929L10.7071 7.29289C10.3166 6.90237 9.68342 6.90237 9.29289 7.29289Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </li>
        <li
          class=""
        >
          <a
            class="css-lr01fi"
            href="/project/yvzb2a5aaceuaa8vs9kawruc"
            style="color: rgba(255, 255, 255, 0.69);"
          >
            frosty-hamilton
          </a>
        </li>
      </ol>
    </nav>
    <button
      class="css-vl70gq"
    >
      Run
    </button>
  </div>
</div>
`;
